### ISSUE TITLE
Issue trying to export an image used in an extension

### ISSUE THREAD

**Issue by jjcasmar** [KhronosGroup/glTF-Blender-IO#845]
I am trying to create an extension for supporting some very simple materials like spheerical mapping. The material has some parameters and an image and in Blender is implemented a ShaderNodeCustomGroup. 

For exporting this material, I am looking for the socket where the ImageTexture node is connected and I use the `gather_texture_info` function which is used for other textures in the standard material. This function returns a `gltf2_io.TextureInfo`

However, when I add this object to my extension definition, I get an error "TextureInfo is not JSON serializable" although is used several times accross the code. 

```
texture_info = self.gather_texture_info((env_socket,), export_settings)
gltf2_material.extensions[bl_info['name']] = self.Extension(
      name=bl_info['name'],
      extension={"img": texture_info},
      required=False
)
```

How should I add the texture info so It works?

**Comment by scurest** [comment]
Turn it into a dict with `.to_dict()`.

**Comment by jjcasmar** [comment]
Shouldn;t that be called automatically when trying to serialize the TextureInfo? For example, in gltf_io.py, for the baseColorTexture there is

```
result["baseColorTexture"] = from_union([lambda x: to_class(TextureInfo, x), from_none],
                                                self.base_color_texture)
```
which somehow calls the appropiate stuff. 

I was expecting this happens automatically given my `TextureInfo` is in the extensions object...

**Comment by jjcasmar** [comment]
In fact, doing `to_dict()` gives the following error:

```
  File "/home/kdab/.config/blender/2.81/scripts/addons/kuesa/materials/KDAB_custom_materials.py", line 139, in gather_material_hook
    extension={"img": texture_info.to_dict()},
  File "/home/kdab/.config/blender/2.81/scripts/addons/io_scene_gltf2/io/com/gltf2_io.py", line 661, in to_dict
    result["index"] = from_int(self.index)
  File "/home/kdab/.config/blender/2.81/scripts/addons/io_scene_gltf2/io/com/gltf2_io.py", line 29, in from_int
    assert isinstance(x, int) and not isinstance(x, bool)
AssertionError
```

**Comment by scurest** [comment]
Nothing actually gets traversed automatically in the `to_dict` pass. Every single class makes a decision about what to call on each of its fields. Extensions are just dicts and nothing gets called on their fields.

The problem with calling `to_dict` early here is there's a `__traverse` pass that gathers stuff like accessors into a top-level array and replaces them with an index, and then there's a `to_dict` pass that turns the classes into dicts, and since the `to_dict` pass is second it assumes everything's been replaced by an index. But extensions want to run `to_dict` early. None of the existing extensions reference any of the top-level arrays so this hasn't been a problem so far.

The easiest immediate fix is for you to just convert the TextureInfo into a dict yourself. I tested this and it works fine.

I guess `to_dict` could be changed to traverse extensions and extras and call `to_dict` on stuff that isn't valid JSON (sort of like `__to_json_compatible`).

**Comment by jjcasmar** [comment]
> Nothing actually gets traversed automatically in the `to_dict` pass. Every single class makes a decision about what to call on each of its fields. That line you quoted is explicitly calling `to_dict` (through `to_class`) on the TextureInfo object.
> 
> The problem with calling `to_dict` early here is there's a `__traverse` pass that gathers stuff like accessors into a top-level array and replaces them with an index, and then there's a `to_dict` pass that turns the classes into dicts, and since the `to_dict` pass is second it assumes everything's been replaced by an index (hence the assert). But extensions want to run `to_dict` early. None of the existing extensions reference any of the top-level arrays so this hasn't been a problem so far.

I have another extension which creates a top level entry of layers and then the nodes reference the layers where they belong. This works fine, although probably because the layer is just a string, with is serializable. 

```
            layers = [self.current_collections[collection.name] for collection in blender_object.users_collection]
            gltf2_object.extensions[bl_info['name']] = self.Extension(
                name=bl_info['name'],
                extension={"layers": layers},
                required=False
            )
```

> 
> The easiest immediate fix is for you to just convert the TextureInfo into a dict yourself. I tested this and it seems to work fine.
> 
> I guess `to_dict` could be changed to traverse extensions and extras and call `to_dict` on stuff that isn't valid JSON (sort of like `__to_json_compatible` does).

I like the idea of being able to call recursively to_dict for extensions or something like that. Can we discuss that?

**Comment by scurest** [comment]
[Proof of concept](https://github.com/KhronosGroup/glTF-Blender-IO/compare/master...scurest:jsonify-exts-and-extras) for traversing extensions and extras and calling `to_dict` and `to_list` on stuff. This makes your original code work.

Let's see if anyone has other ideas. The official addon will also have to solve this somehow if it ever supports exporting pbrSpecularGlossiness materials.

**Comment by jjcasmar** [comment]
I have tried the code and works fine, but I have some doubts. In gltf2_io, for the `baseColorTexture` data, you use 

```
result["baseColorTexture"] = from_union([lambda x: to_class(TextureInfo, x), from_none], self.base_color_texture)
```
`self.base_color_texture` is a `TextureInfo` and in this case we use this line of code to convert the `TextureInfo` to some serializable while if using a `TextureInfo` in an extension, the `to_dict()` code will be called (through the `jsonify`). Is this going to cause some kind of trouble?

**Comment by scurest** [comment]
What sort of trouble? That code just checks that self.base_color_texture is either an instance of TextureInfo, in which case `.to_dict()` is called on it, or None. IOW that line "knows" what the type of self.base_color_texture should be and will throw an error if its wrong. But for extensions and extras we don't know what the type of each property should be.

**Comment by jjcasmar** [comment]
Understood. My Python knowledge is still limited :)
The code works fine, are you going to create a PR with it?

**Comment by scurest** [comment]
(It is a rather opaque line.) Nah, I wanna see what other people think.

**Comment by jjcasmar** [comment]
Do you think its possible to try to get this into 2.82? Without this, I think user extensions are not so useful

**Comment by jjcasmar** [comment]
The json encoder support a default function for non serializable object. Maybe we can use this? I have done some tests passing the following function
`lambda x: x.to_dict()`
and works more or less correctly. The only issue is that it adds some data that shouldn't be there. 
```                    "img" : {
                        "extensions" : {},
                        "extras" : null,
                        "index" : 0,
                        "texCoord" : 0
                    }
```

If there is an easy way to remove the extras and extensions if not needed, maybe its an easier way and less intrusive for the rest of the code. 

What do you think?

**Comment by scurest** [comment]
That sounds like a good idea. Do the extensions and extras properties not get removed by [`__fix_json`](https://github.com/KhronosGroup/glTF-Blender-IO/blob/331776410f1d3feb55d19b74e8f230166aef27f0/addons/io_scene_gltf2/blender/exp/gltf2_blender_export.py#L88) already?

**Comment by jjcasmar** [comment]
Im trying to find why is not happening

**Comment by scurest** [comment]
Oh, `__fix_json` is called before `json.dump`.

**Comment by jjcasmar** [comment]
Im trying to understand what is happening here.
So the exporter produces a dictionary. For the extension part, the dictionary basically contains "{img: instance of TextureInfo}".
Then we do the `__fix_json` stuff, which of course does nothing for the TextureInfo object.
Finally, we call dumps which produces the json string to be saved. At this point is where the `to_dict()` is called. 

Moving to a custom encoder means that we can remove the not needed keys while encoding, right?

**Comment by scurest** [comment]
Actually, would it be easier to add the `to_dict()` stuff in `__fix_json`?

**Comment by jjcasmar** [comment]
I have done something similar. Works fine, but seems a bit out of place.

**Comment by scurest** [comment]
Isn't that what my first proof of concept did? You could write it like this instead

````py
result["extensions"] = from_union([lambda x: from_dict(lambda x: from_dict(jsonify, x), x), from_none],
                                  self.extensions)
````

**Comment by jjcasmar** [comment]
Right. I;m not sure which one looks better. One is a oneliner, but looks a bit out of place. The other one is more complex, but the result is a directly serializable dictionary. 

Are extensions of extensions allowed? If they are allowed, I think the only valid method is the oneliner, as its recursive and will produce the correct result for the inner extension. I think your first POC will fail in this case, no?

**Comment by scurest** [comment]
Extensions of extensions are allowed. The first PoC is recursive too, it works fine with nested extensions.

I also realized that one thing just changing `__fix_json` can't handle is

```
'extensions': {
    'my_ext': TextureInfo instance
}
```

(ie. the extension value is a TextureInfo instance directly, not a `{'img': TextureInfo instance}`) because `to_dict` demands that the values in the extensions dict be dicts themselves. Fixing this in `__fix_json` is too late, `to_dict` has already been called and asserted that the values are dicts.

So I think it is better to change `to_dict` like in the first PoC.

**Comment by jjcasmar** [comment]
Im not sure I understand the problem

**Comment by scurest** [comment]
If you only change `__fix_json` this is ok

````py
gltf2_material.extensions[bl_info['name']] = self.Extension(
      name=bl_info['name'],
      extension={"img": texture_info},
      required=False
)
````

but this doesn't work

````py
gltf2_material.extensions[bl_info['name']] = self.Extension(
      name=bl_info['name'],
      extension=texture_info,
      required=False
)
````

**Comment by jjcasmar** [comment]
Yep, you are right. The issue is in `from_dict`, right? Because we expect the extensions to be a dictionary and in this case it isnt.

**Comment by scurest** [comment]
In `to_dict`, yeah.

**Comment by jjcasmar** [comment]
I am trying a recursive extension and having some issues with your first POC. 

```
texture_info = self.gather_texture_info((env_socket,), export_settings)
                    texture_info.extensions = {"img"}


                    gltf2_material.extensions[bl_info['name']] = self.Extension(
                        name=bl_info['name'],
                        extension=texture_info,
                        required=False
                    )
```

**Comment by scurest** [comment]
Thinking again I'm good with your current PR. It's should be easy to get merged and it will be fine for current extensions.

If anyone ever does want a non-dict as an extension value we can revisit with a more robust implementation then. Existing code won't have to change.

**Comment by jjcasmar** [comment]
Are you sure about it? I expect people raising issues when trying to create their own extensions which we wont be aware of...

**Comment by scurest** [comment]
Well, if you do want to change it, that's fine with me too.

**Comment by jjcasmar** [comment]
I will close the PR for now and I will reopen it tomorrow with your POC. Or maybe its better if you create it, since its your code :)

**Comment by scurest** [comment]
Not too sure about the details of my PoC. And I'm still trying to annoy someone into noticing my existing PRs lol ( ´ ▽ ` ;) If you want to pull it into your own branch, go for it.

**Comment by jjcasmar** [comment]
I see a big caveat in using something like your POC. That file is automatically generated by quicktype, so we cant modify it.

**Comment by scurest** [comment]
You're right. Could be done as a separate pass just before `to_dict` then but like I said, I wanna see if anyone has better ideas.

**Comment by jjcasmar** [comment]
I have been checking and that file has been modified manually several times, so I guess there's no problem

**Comment by scurest** [comment]
@emackey This was closed by #849.

**Comment by emackey** [comment]
Thanks @scurest and @jjcasmar
